File: scheduler.py
import datetime

from logger import logger

TZ_OFFSET = datetime.timezone(datetime.timedelta(hours=3))


def get_now():
    return datetime.datetime.now(TZ_OFFSET)


def calculate_next_time(after_time, config):
    if after_time.tzinfo is None:
        after_time = after_time.replace(tzinfo=TZ_OFFSET)

    slots = []
    for s in config['preferred_slots']:
        h, m = map(int, s.split(':'))
        slots.append(after_time.replace(hour=h, minute=m, second=0, microsecond=0))
    slots.sort()

    for slot in slots:

        if slot >= after_time + datetime.timedelta(seconds=config['min_interval']):
            if config['window_start'] <= slot.hour < config['window_end']:
                return slot

    tomorrow_morning = (after_time + datetime.timedelta(days=1)).replace(
        hour=config['window_start'], minute=0, second=0, microsecond=0
    )

    return calculate_next_time(tomorrow_morning - datetime.timedelta(seconds=config['min_interval']), config)


def rebalance_queue(db):
    logger.info("Rebalance queue")
    config = db.get_config("scheduler")
    if not config:
        logger.info("Not config")
        return

    logger.info("Config ok")
    now = get_now()
    last_publish_us = db.get_latest_published_time()
    day_stats = {}

    if last_publish_us:
        last_dt = datetime.datetime.fromtimestamp(last_publish_us / 1000000, tz=TZ_OFFSET)
        current_point = max(now, last_dt)
        if last_dt.date() == now.date():
            day_stats[now.date()] = 1
    else:
        current_point = now

    logger.info("Get posts")

    scheduled_posts = sorted(
        db.get_scheduled_queue(),
        key=lambda post: print(post.created_at, type(post.created_at)) or post.created_at
    )
    print(scheduled_posts)

    updates = []
    logger.info("Start scheduling")

    for post in scheduled_posts:
        while True:
            next_time = calculate_next_time(current_point, config)
            post_date = next_time.date()
            current_day_count = day_stats.get(post_date, 0)

            if current_day_count < config['posts_per_day']:
                ts = int(next_time.timestamp() * 1000000)

                updates.append({
                    "id": post.id,
                    "publish_at": ts
                })

                day_stats[post_date] = current_day_count + 1
                current_point = next_time
                break
            else:
                current_point = next_time.replace(hour=23, minute=59, second=59)

    if updates:
        db.update_posts_batch(updates)
        logger.info(f"Batch updated {len(updates)} posts in queue.")

File: cf/src/scheduler.ts
import type { PostRow, SchedulerConfig } from './types'

export const TZ_OFFSET_MS = 3 * 3600_000

export interface QueueDb {
    getConfig(key: 'scheduler'): Promise<SchedulerConfig | null>
    getLatestPublishedMs(): Promise<number | null>
    getScheduledQueue(): Promise<PostRow[]>
    updatePostsBatch(updates: QueueUpdate[]): Promise<void>
}

interface LocalParts {
    y: number
    m: number
    d: number
    h: number
    min: number
    s: number
}

function localParts(ms: number): LocalParts {
    const d = new Date(ms + TZ_OFFSET_MS)
    return {
        y: d.getUTCFullYear(),
        m: d.getUTCMonth(),
        d: d.getUTCDate(),
        h: d.getUTCHours(),
        min: d.getUTCMinutes(),
        s: d.getUTCSeconds(),
    }
}

function makeLocal(y: number, m: number, d: number, h: number, min: number, s: number): number {
    return Date.UTC(y, m, d, h, min, s) - TZ_OFFSET_MS
}

export function calculateNextTime(afterMs: number, config: SchedulerConfig): number {
    let iter = 0
    let current = afterMs
    for (;;) {
        if (++iter > 1000) throw new Error(`calculateNextTime infinite loop afterMs=${afterMs} config=${JSON.stringify(config)}`)
        const { y, m, d } = localParts(current)

        const slots: number[] = []
        for (const slot of config.preferred_slots) {
            const [h, min] = slot.split(':').map(Number)
            if (config.window_start <= h && h < config.window_end) {
                slots.push(makeLocal(y, m, d, h, min, 0))
            }
        }
        slots.sort((a, b) => a - b)

        if (slots.length === 0) {
            throw new Error('Ни один preferred_slots не попадает в окно window_start–window_end')
        }

        const found = slots.find((slot) => slot >= current + config.min_interval * 1000)
        if (found !== undefined) return found

        const nextDayStart = makeLocal(y, m, d + 1, config.window_start, 0, 0)
        const candidate = nextDayStart - config.min_interval * 1000
        current = dayKey(candidate) === dayKey(nextDayStart) ? candidate : nextDayStart
    }
}

function dayKey(ms: number): number {
    return Math.floor((ms + TZ_OFFSET_MS) / 86_400_000)
}

export interface QueueUpdate {
    id: string
    publish_at: number
    sequence_number: number
}

export async function rebalanceQueue(db: QueueDb, nowMs: number = Date.now()): Promise<void> {
    const config = await db.getConfig('scheduler')
    if (!config) return

    const lastPublishMs = await db.getLatestPublishedMs()
    const dayStats = new Map<number, number>()

    let currentPoint: number
    if (lastPublishMs !== null) {
        currentPoint = Math.max(nowMs, lastPublishMs)
        if (dayKey(lastPublishMs) === dayKey(nowMs)) {
            dayStats.set(dayKey(nowMs), 1)
        }
    } else {
        currentPoint = nowMs
    }

    const scheduledPosts = [...(await db.getScheduledQueue())].sort(
        (a, b) => a.created_at - b.created_at
    )

    const updates: QueueUpdate[] = []

    for (const post of scheduledPosts) {
        for (;;) {
            const nextTime = calculateNextTime(currentPoint, config)
            const key = dayKey(nextTime)
            const count = dayStats.get(key) ?? 0

            if (count < config.posts_per_day) {
                updates.push({
                    id: post.id,
                    publish_at: nextTime,
                    sequence_number: updates.length + 1,
                })
                dayStats.set(key, count + 1)
                currentPoint = nextTime
                break
            }
            // Дневной лимит исчерпан — прыгаем на конец локального дня и ищем слот заново
            const p = localParts(nextTime)
            currentPoint = makeLocal(p.y, p.m, p.d, 23, 59, 59)
        }
    }

    if (updates.length > 0) {
        await db.updatePostsBatch(updates)
    }
}
